### PR TITLE
Moved floodFill algorithms to algorithm extra

### PR DESCRIPTION
**Pull request by bennlich**
Not sure what else will go here, but presumably there will be things.

Also, currently I'm modifying class prototypes, which is probably fine, but eventually it might be nice to use a syntax like

    extend(ABM.AgentSet, {
       newProp1: blah
       newProp2: blah
    })

etc.

### CODE REVIEW

**Comment by bennlich** [comment]
I've started using this FloodFill class in seismogram and it seems to be satisfying my needs. Wuddya think @backspaces ?

**Comment by bennlich** [comment]
Another example of what this allows for/simplifies: floods with dynamic candidate/join functions. In seismogram I wrap ABM.FloodFill in an object that keeps track of a constantly-changing threshold used by the floods candidate function.

**Comment by backspaces** [comment]
It's fascinating: using a class that creates FF instances!  Er.. right?  :)
 Way cool, and love the encapsulation.

Speaking of which, I think you can remove the three function args to
floodFillOnce: the three functions are already instance variables,
according to the ctor.

  constructor: (@startingSet, @fCandidate, @fJoin, @fNeighbors) ->
    @nextStep = () => @floodFillOnce @startingSet, @fCandidate,
@fJoin, @fNeighbors
  floodFillOnce: (aset, fCandidate, fJoin, fNeighbors, asetLast=[]) =>

I.e. floodFillOnce: (aset, asetLast), 'cause fCandidate, fJoin, fNeighbors
are now instance variables and can just be called via @fCandidate etc.

I agree the minor question is whether to make it core, or an extra/? I
haven't a concern either way. Lets go with extras/ unless it creates
awkwardness like dataset has specifying the model in patch variables etc.
(We'll simplify that later)

The patch subclass is fascinating too, but documentation would fill the
bill also, as I think the only difference is the defaulted fNeighbors ..
which is cool, I gotta say.  But if we do it for patches, we may also want
to do it for links using the "linkNeighbors". Or outLinks/inLinks.

Again, the documentation (docco) can show it, and using it in existing
models is a good idea.  I could probably re-write the buttons model to use
it to find all the subgraphs, for example .. gotta check.

But definitely check the dude in, its clean clear and useful .. and a great
solution to our earlier issues!

**Comment by bennlich** [comment]
Hey! Cool idea. Now that you mention it. I think we can get rid of all the arguments to `floodFillOnce()`. How's this:
```
class ABM.FloodFill
  constructor: (startingSet, @fCandidate, @fJoin, @fNeighbors) ->
    @nextStep = () => @floodFillOnce()
    @nextFront = startingSet
    @prevFront = []

  floodFillOnce: () =>
    @fJoin p, @prevFront for p in @nextFront
    asetNext = []
    for p in @nextFront
      for n in @fNeighbors(p) when @fCandidate n, @nextFront
        asetNext.push n if asetNext.indexOf(n) < 0
    
    @prevFront = @nextFront
    @nextFront = asetNext

    if @nextFront.length is 0
      @nextStep = null
```

Could even get rid of `nextStep()` now, but I think it's useful for knowing when the flood is complete? Or we could instead pass in an `onComplete()` callback.

**Comment by bennlich** [comment]
One more question: do you think flooding over links should be called `ABM.Links.FloodFill` or `ABM.Agents.FloodFill`? And should you start it off by passing in a node (agent) or a link?

**Comment by backspaces** [comment]
Man, that *does* reduce/simplify a lot, nice catch.

Re: getting rid of nextStep
- Is there an advantage to having floodFillOnce setting/returning an @done bool, say, rather than setting the nextStep to null? I think setting nextStep to null is fine too, as currently done, especially if you find it OK for your current nifty stunts for dynamic usage.
- I really think it would be useful to have a simple floodFill method that does all the steps for you, so you don't need to loop or manage floodFillOnce.  I think that's possible, right? I typically don't need to "be in the loop" so to speak, just let the flood complete without intervention.

In terms of links, patches etc: I'd prefer the basic class as an extra initially if there is nothing awkward about using it, removing the existing one from Patches etc.  Documentation & demos can help a lot.  And generality is hard: sometimes n4 is used for the neighborhood or just the out-links etc.  Lets start slow and steady with a simple nifty dude and see if it is not too awkward.  I'll try it in buttons if I'm right about its use finding connected subgraphs, which'd be cool!  If not I can just build a simple sketch for flooding a graph.

**Comment by bennlich** [comment]
I like the `@done` bool idea. I think it's clearer than setting nextStep to null. And I agree, being able to do a complete flood all at once is key. We could either have a function that instantiates and runs a flood fill object:
```
ABM.Patches.floodFill: (args) =>
  flood = new ABM.Patches.FloodFill(args)
  flood.nextStep() while not flood.done
```
or we could add a `go()` function to the class:
```
class ABM.FloodFill
  constructor: (startingSet, @fCandidate, @fJoin, @fNeighbors) ->
    @nextFront = startingSet
    @prevFront = []
    @done = false

  nextStep: () =>
    if @done return

    @fJoin p, @prevFront for p in @nextFront
    asetNext = []
    for p in @nextFront
      for n in @fNeighbors(p) when @fCandidate n, @nextFront
        asetNext.push n if asetNext.indexOf(n) < 0
    
    @prevFront = @nextFront
    @nextFront = asetNext

    if @nextFront.length is 0
      @done = true
  
  go: () =>
    @nextStep() while not @done
```
and then you'd do `new ABM.Patches.FloodFill(args).go()` instead of `ABM.Patches.floodFill(args)`.

Wuddya think?

**Comment by backspaces** [comment]
God yes, the go() idea is great. Nice!

Sweet.

**Comment by bennlich** [comment]
You prefer go() to wrapping the object in another function?

**Comment by backspaces** [comment]
Just to make sure, I'd prefer the .go() example below, using a FloodFill class, if I'm understanding correctly.  Its kinda nice to hint that there is a single step approach too to a modeler who wasn't aware of that feature.

    new ABM.Patches.FloodFill(args).go() instead of ABM.Patches.floodFill(args).

But I'd prefer to *not* have separate versions for Patches & Agents (using either radius or links, say), just an extra/ that creates ABM.FloodFill that is not dependent on its environment.  I think that's possible, right? Or is there a name-space issue I'm not understanding?

Nit: do you need the =>'s or would plain ->'s work?  I doubt it makes a difference but just curious, .. could be a name-space problem I'm not aware of.

**Comment by bennlich** [comment]
Okay. I like the idea of hinting at the single step approach with go().

Why no subclasses for Patches/Links?

Single-arrows are probably fine. I think I was just using double-arrows out of habit.

**Comment by floodfill** [comment]
Hey don't cc me on your email threads

Sent from Kan's fake iPhone

**Comment by backspaces** [comment]
Why no subclasses for Patches/Links?

Mainly because I don't understand them!  :)

Is it simply to help with defaults for the fNeighbors?  I.e. (p) -> p.n4 or n?  Or is there a name-scope issue? Or for graphs, I guess it could be links, outLinks or inLinks, right? 

I guess my "principles" are to make the agentsets fairly basic, just enough to manage the NL functions like nOf, oneOf, etc. FloodFill seems so general and powerful, I'm not sure I could provide the right defaults for fNeighbors.

Lets try a talky, hangout or similar .. and you'll be in town soon so maybe at the office.

   -- Owen

**Comment by bennlich** [comment]
> Is it simply to help with defaults for the fNeighbors? I.e. (p) -> p.n4 or n?

Yeah, I always thought having patches.floodFill (in addition to the more general agentset.floodFill) was a nice convenience, though I could also get on board with keeping things minimal / simple. Maybe it would be enough to include examples of floodFilling patches and links in a comment and/or in example models?

**Comment by stephenguerin** [comment]
>
> Benny writes
>
> Yeah, I always thought having patches.floodFill (in addition to the more
> general agentset.floodFill) was a nice convenience, though I could also get
> on board with keeping things minimal / simple. Maybe it would be enough to
> include examples of floodFilling patches and links in a comment and/or in
> example models?
>
>
>
I lean toward floodfill as more of a modeling pattern and not a framework
requirement. In October I wrote:

I talked to Owen about this last month. A flood fill is a general modeling
> pattern that can become specialized (eg spreading activation, graph search,
> percolation, etc) with a variety of implementation needs. I would say it
> probably should be the responsibility of the modeler and not the framework.

**Comment by floodfill** [comment]
Hey why I'm cc'ed in your threads?

Sent from Kan's fake iPhone

**Comment by stephenguerin** [comment]
Kam, did you subscribe to the agentscript github repository? We are not
explicitly cc'ing you

**Comment by floodfill** [comment]
Nope I didn't.

Sent from Kan's fake iPhone

> 在 2015年3月2日，13:34，Stephen Guerin <notifications@github.com> 写道：
> 
> Kam, did you subscribe to the agentscript github repository? We are not
> explicitly cc'ing you

**Comment by bennlich** [comment]
Oh, rofl. I think it's because we accidentally tagged your username when talking about a coffeescript variable called "floodfill". I've never seen this happen before. Let me see if I can un-tag you by editing that post (seems unlikely, though). Might have to e-mail support.

**Comment by bennlich** [comment]
Kan, do you see an "unsubscribe" button anywhere on [this page](https://github.com/backspaces/agentscript/pull/75). It should look like the one described [here](https://help.github.com/articles/unsubscribing-from-conversations/).

**Comment by stephenguerin** [comment]
aya,  sorry you got caught in the thread, Kan.

BTW, I'm curious how you picked the username "floodfill".

--- -. .   ..-. .. ... ....   - .-- ---   ..-. .. ... ....
Stephen.Guerin@Redfish.com
1600 Lena St #D1, Santa Fe, NM 87505
office: (505) 995-0206  mobile: (505) 577-5828
tw: @redfishgroup  skype: redfishgroup
redfish.com  |  simtable.com

**Comment by backspaces** [comment]
Kewl!

**Comment by bennlich** [comment]
Hmm. I just found a bug in my fJoin / fCandidate functions. If you start two floods at once, I realized you have to flag flooded patches that satisfy the fCandidate function to prevent them from being candidates for (and thus joining) more than one flood.

This made me wonder about having both fJoin and fCandidate. Is it necessary to have two functions? Or does it suffice to have one that tests for candidacy *and* adds the new patch to some list / marks it with a flood id?

**Comment by backspaces** [comment]
God. Do you mean in a multi-model situation, or in an async situation where
two floods could occur at the same time within the same model?

I'd prefer to not to use marks in the patches, but there are advantages.
We currently use indexOf, as I recall, to see if we've visited a patch.

Would it solve the problem to merge join/candidate?  Good example of
discovering a better way after refactoring, like in the class
implementation.  If it works, it would be nice to simplify.

I hope to make wedtech tomorrow, maybe we can chat about it.

   -- Owen

**Comment by bennlich** [comment]
Nothing to do with multi-model--just an async situation where two floods progress simultaneously. Imagine you have a situation like this:

A 0 B
A 0 B
A 0 B

where A's are patches flooded by flood A, B's are patches flooded by flood B, and 0's are patches yet-to-be-flooded. At this time step, both flood A and flood B are going to test the 0's, and unless one of the two somehow marks the 0's as "claimed" during fCandidate, the 0's will pass the candidacy test for both floods, passing them along to nextStep to be joined.

What if we eliminated fJoin altogether? So if previously you had something like:
```coffee
fJoin = (p) ->
    p.floodDistance = @steps
    p.floodId = @id
    p.color = @color
    model.floods[@id].push(p)

fCandidate = (p) ->
    passesTheTest =
        p.brightness > someValue
        && (typeof p.floodId === 'undefined') // patch has not already been flooded

    return passesTheTest
```

Instead you'd combine the two into something like:

```coffee
fTest = (p) ->
    passesTheTest =
        p.brightness > someValue
        && (typeof p.floodId === 'undefined') // patch has not already been flooded
    
    if (passesTheTest)
        p.floodDistance = @steps
        p.floodId = @id
        p.color = @color
        model.floods[@id].push(p)

    return passesTheTest
```

**Comment by backspaces** [comment]
God, I can't believe I didn't catch this earlier.

TL;DR: During dynamic computations, "marking" is a bad idea, better
replaced by an array instead. I think!

I've always been suspicious of "marking" a patch in floods and diffusions.
I think diffusions have a similar issue, having to have memory used by its
two passes.

For diffusions, I had recently planned to add them to DataSets.  Diffusing
over a dataset would use a second dataset. One of the two datasets would be
GC'd when done, depending on the algorithm. Since it is a single large
object, GC is not costly.

This would have two advantages
1 - Making diffusion an algorithm within DataSets, way cool and useful!
2 - Preventing expensive temporary variables within Patch/Agent instances

I think we may want this for floods: an array which collects the results of
the flood. This could be done now by the modeler, having the two async
floods each keeping an array of flood members.  No marking. This would be
too awkward for the modeler, best done by FloodFill itself.

This is a general issue that es7 is likely to expose: Using variables which
are static, during a dynamic computation, is Considered Harmful! This is
why array comprehensions were removed from es6: they hadn't resolved
parallelism for comprehensions.

I can think of situations where "marking" a flood is not needed, the flood
is confined to a step.  For example, using a flood to answer "can I get
there from here" (gridpath model) currently marks.  In a huge model, that
would be costly and an array kept by step would make sense. Now the
floodfill would "return" a value too.

Would this help in the A 0 B situation?

I'm not hugely concerned about marking if it really is leaving a useful
piece of info within the patch that will last over the life of the model.
But we may want a way to use an array which is returned by the flood in the
case of the flood leaving a temporary variable.
